File: webcam/webcam.cpp
/* 2010-10-28
   C. Schmidt-Hieber, University College London */

#include <cstdio>
#include <iostream>
#include <iomanip>
#include <ctime>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/types.h>
#include <fcntl.h>
#include <sstream>
#include <algorithm>
#include <opencv2/opencv.hpp>

static const double BILLION = 1000000000L;
static const int BUFSIZE = 4096;
static const int WIDTH = 320;
static const int HEIGHT = 240;
static const int FPS = 30;
static const int CAM_ID = 2;
static const bool g_rotate = false;

double tdiff(timespec time1, timespec time0) {
    return ( time1.tv_sec - time0.tv_sec )
         + ( time1.tv_nsec - time0.tv_nsec ) / BILLION;
}

double t2d(timespec time1) {
    return time1.tv_sec + time1.tv_nsec / BILLION;
}

void write_send(int socket, const cv::Mat& im, CvVideoWriter* pWriter, int& ncount) {
    timespec time1;
    if( clock_gettime( CLOCK_REALTIME, &time1) == -1 ) {
        fprintf(stderr, "WEBCAM: clock gettime");
    }
    double dtime1 = t2d(time1);
    
    while (send(socket, &dtime1, sizeof(double), 0) < 0) {
        perror("WEBCAM: client: send");
    }
    IplImage ipl_im(im);
    cvWriteFrame(pWriter, &ipl_im);
    ncount++;
}

void get_image(cv::VideoCapture& camera, cv::Mat& im, const cv::Mat& mapping, bool rotate, int socket, CvVideoWriter* pWriter, int& ncount) {
    camera >> im;
    cv::Mat cpim = im.clone();
    if (rotate) {
        cv::warpAffine(cpim, im, mapping, im.size());
    }
    if (pWriter != NULL) {
        write_send(socket, im, pWriter, ncount);
    }
}

int main (int argc, char **argv)
{
    char* ntry = (char*)"";
    if (argc > 1) {
        ntry = argv[1];
    }

    double fps = FPS;
    double target_dur = 1.0/fps;
    double tol = 1.0e-3;
    
    cv::VideoCapture camera(CAM_ID);
    if (camera.isOpened()) {
        camera.release();
    }
    camera.open(CAM_ID);

    camera.set(CV_CAP_PROP_FRAME_WIDTH, WIDTH);
    camera.set(CV_CAP_PROP_FRAME_HEIGHT, HEIGHT);
    camera.set(CV_CAP_PROP_FPS, FPS);
    
    int width = camera.get(CV_CAP_PROP_FRAME_WIDTH);
    int height = camera.get(CV_CAP_PROP_FRAME_HEIGHT);
    std::cout << "WEBCAM: Initialized camera with " << width << "x" << height << std::endl;
#ifdef STANDALONE
    std::cout << "WEBCAM: fps " << camera.get(CV_CAP_PROP_FPS) << std::endl;
    std::cout << "WEBCAM: fourcc " << camera.get(CV_CAP_PROP_FOURCC) << std::endl;
    std::cout << "WEBCAM: Format " << camera.get(CV_CAP_PROP_FORMAT) << std::endl;
    std::cout << "WEBCAM: Mode " << camera.get(CV_CAP_PROP_MODE) << std::endl;
    std::cout << "WEBCAM: Brightness " << camera.get(CV_CAP_PROP_BRIGHTNESS) << std::endl;
    std::cout << "WEBCAM: Contrast " << camera.get(CV_CAP_PROP_CONTRAST) << std::endl;
    std::cout << "WEBCAM: Saturation " << camera.get(CV_CAP_PROP_SATURATION) << std::endl;
    std::cout << "WEBCAM: Hue " << camera.get(CV_CAP_PROP_HUE) << std::endl;
    std::cout << "WEBCAM: Gain " << camera.get(CV_CAP_PROP_GAIN) << std::endl;
    std::cout << "WEBCAM: Exposure " << camera.get(CV_CAP_PROP_EXPOSURE) << std::endl;
    std::cout << "WEBCAM: Convert to RGB " << camera.get(CV_CAP_PROP_CONVERT_RGB) << std::endl;
    // std::cout << "WEBCAM: White Balance " << camera.get(CV_CAP_PROP_WHITE_BALANCE) << std::endl;
#endif
    cv::Mat mapping = cv::getRotationMatrix2D(cv::Point2f(width/2.0, height/2.0), 180.0, 1.0);
    cv::namedWindow("Webcam");
    cvMoveWindow("Webcam", 1280-WIDTH, 0);
    cv::Mat im;
    int ncount = 0;
    get_image(camera, im, mapping, g_rotate, -1, false, ncount);
    cv::imshow("Webcam", im);
    
#ifndef STANDALONE
    int s;
    if ((s = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
        perror("WEBCAM: server: socket");
        return -1;
    }

    /*
     * Create the address we will be connecting to.
     */
    sockaddr_un sa;
    sa.sun_family = AF_UNIX;

    std::ostringstream tmpfn;
    tmpfn << "vidsocket" << ntry;
    std::cout << "WEBCAM: socket name " << tmpfn.str() << std::endl;

    int nameLen = strlen(tmpfn.str().c_str());
    if (nameLen >= (int) sizeof(sa.sun_path) -1) { /* too long? */
        return -1;
    }
    
    sa.sun_path[0] = '\0';  /* abstract namespace */
    strcpy(sa.sun_path+1, tmpfn.str().c_str());
    int len = 1 + nameLen + offsetof(struct sockaddr_un, sun_path);

    // sockaddr_un sa;

    // std::string tmpfn("vidsocket");
    
    // int nameLen = strlen(tmpfn.c_str());
    // if (nameLen >= (int) sizeof(sa.sun_path) -1)  /* too long? */
    //     return -1;

    // sa.sun_family = AF_UNIX;
    // sa.sun_path[0] = '\0';  /* abstract namespace */
    // strcpy(sa.sun_path+1, tmpfn.c_str());

    /*
     * Try to connect to the address.  For this to
     * succeed, the server must already have bound
     * this address, and must have issued a listen()
     * request.
     *
     * The third argument indicates the "length" of
     * the structure, not just the length of the
     * socket name.
     */

    if (connect(s, (sockaddr*)&sa, len) < 0) {
        perror("WEBCAM: client: connect");
        return -1;
    }

    bool connected = false;
    std::vector<char> data(BUFSIZE);
    int nrec = recv(s, &data[0], data.size(), 0);
    std::string datastr(data.begin(), data.end());
    if (nrec<=0) {
        std::cerr << "WEBCAM: Didn't receive start message; exiting now" << std::endl;
        return -1;
    }
    connected = true;
    
    std::string ready = "ready";
    while (send(s, ready.c_str(), ready.size(), 0) < 0) {
        perror("WEBCAM: client: send");
    }

    int flags;
    if (-1 == (flags = fcntl(s, F_GETFL, 0)))
        flags = 0;

    if (fcntl(s, F_SETFL, flags | O_NONBLOCK)==-1) {
        perror("WEBCAM: client: unblock");
    }
#endif

    CvVideoWriter* pWriter = NULL;
    
    timespec time0;
    if( clock_gettime( CLOCK_REALTIME, &time0) == -1 ) {
        fprintf(stderr, "WEBCAM: clock gettime");
    }
    timespec time1 = time0;
    timespec time2 = time0;
    timespec time3 = time0;
    timespec t_disconnect = time0;
    int nframes = 0;

#ifdef STANDALONE
    int s = -1;
    std::string fn = "";
#else
    std::string fn = "";
#endif
    while (true) {
        if( clock_gettime( CLOCK_REALTIME, &time1) == -1 ) {
            fprintf(stderr, "WEBCAM: clock gettime");
        }
#ifndef STANDALONE
        std::vector<char> data(BUFSIZE, '\0');
        int nrec = recv(s, &data[0], data.size(), 0);
        std::string datastr(data.begin(), data.end());
#endif
        get_image(camera, im, mapping, g_rotate, s, pWriter, ncount);

        cv::imshow("Webcam", im);

        if( clock_gettime( CLOCK_REALTIME, &time2) == -1 ) {
            fprintf(stderr, "WEBCAM: clock gettime");
        }
        double capture_dur = tdiff(time2, time1);
        double rem = target_dur - capture_dur;
        if (rem > tol) {
            nframes += 1;
        } else {
            int dropped_frames = int((capture_dur+tol)/target_dur);
            nframes += dropped_frames +1;
            for (int nf = 0; nf < dropped_frames; ++nf) {
                if (pWriter != NULL) {
                    write_send(s, im, pWriter, ncount);
                }
            }
        }
        double next_timestamp = (nframes * target_dur) + t2d(time0);
        
        if( clock_gettime( CLOCK_REALTIME, &time3) == -1 ) {
            fprintf(stderr, "WEBCAM: clock gettime");
        }
        double total_wait = next_timestamp -  t2d(time3);
        while (total_wait < tol) {
            total_wait += target_dur;
            nframes++;
            if (pWriter != NULL) {
                write_send(s, im, pWriter, ncount);
            }
        }

        int key = cv::waitKey(int(1e3*total_wait));

        while (key != -1) {
            next_timestamp = (nframes * target_dur) + t2d(time0);
            if( clock_gettime( CLOCK_REALTIME, &time3) == -1 ) {
                fprintf(stderr, "WEBCAM: clock gettime");
            }
            if (t2d(time3) < next_timestamp) {
                total_wait = next_timestamp - t2d(time3);
            } else {
                if (pWriter != NULL)
                    cvReleaseVideoWriter(&pWriter);
                
                exit(0);
            }
            key = cv::waitKey(int(1e3*total_wait));
        }

#ifndef STANDALONE
        
        // no update from blender in a long time, terminate process
        if (datastr.find("1")==std::string::npos) {
            if (connected) {
                t_disconnect = time1;
                connected = false;
            } else {
                if (tdiff(time1, t_disconnect) > 0.5) {
                    std::cout << "WEBCAM: Received termination signal" << std::endl;
                    if (pWriter != NULL)
                        cvReleaseVideoWriter(&pWriter);
                    return 0;
                }
            }
        } else {
            connected = true;
        }

        // Stop recording
        if (datastr.find("stop") != std::string::npos && pWriter != NULL) {
            cvReleaseVideoWriter(&pWriter);
            pWriter = NULL;
            std::cout << "WEBCAM: Stopping video" << std::endl;
            connected = true;
            ncount = 0;
        }

        // Start recording
        if (datastr.find("avi") != std::string::npos && datastr.find("stop") == std::string::npos && pWriter == NULL) {
            std::size_t startpos = datastr.find("begin")+5; 
            std::size_t endpos = datastr.find("end") - datastr.find("begin") - 5; 
            fn = datastr.substr(startpos, endpos);
            std::cout << "WEBCAM: Starting video, writing to " << fn << std::endl;
            pWriter = cvCreateVideoWriter( fn.c_str(), CV_FOURCC('F', 'M', 'P', '4'), fps, cvSize(width, height),
                                           1);
            connected = true;
            ncount = 0;
        }
#endif
    }
    if (pWriter != NULL)
        cvReleaseVideoWriter(&pWriter);
    return 0;
}

File: py/webcam.py
# 2010-03-19
# C. Schmidt-Hieber, University College London


# Echo client program
import socket
import time
import sys
sys.path.append("/usr/local/lib/python2.6/site-packages")
import subprocess
import cv
import tempfile
import datetime

def get_image(camera, width, height, writer=None, s=None, rotate=True):
    im = cv.QueryFrame(camera)
    imrot = cv.CloneImage(im)
    if rotate:
        cv.WarpAffine(im, imrot, gMapping)
    if writer is not None:
        write_send(camera, imrot, writer, s)
    # im = Image.fromstring("RGB", cv.GetSize(im), im.tostring(), "raw", "BGR")
    return imrot

def write_send(camera, im, writer, s):
    try:
        # print "Sending %f " % time.time()
        s.send(('%f ' % time.time()).encode('latin-1'))
    except:
        pass
    cv.WriteFrame(writer, im)

if __name__ == "__main__":

    fps = 25.0
    target_dur = 1.0/fps

    camera = cv.CreateCameraCapture(0)
    width = cv.GetCaptureProperty(camera, cv.CV_CAP_PROP_FRAME_WIDTH)
    height = cv.GetCaptureProperty(camera, cv.CV_CAP_PROP_FRAME_HEIGHT)
    cv.SetCaptureProperty(camera, cv.CV_CAP_PROP_FRAME_WIDTH, 640)
    cv.SetCaptureProperty(camera, cv.CV_CAP_PROP_FRAME_HEIGHT, 480)
    width = 640
    height = 480
    print("WEBCAM: Initialized camera with", width, "x", height)
    global gMapping
    gMapping = cv.CreateMat(2, 3, cv.CV_32FC1)
    cv.GetRotationMatrix2D((width/2.0,height/2.0), 180.0, 1.0, gMapping)

    cv.NamedWindow("Webcam")
    cv.MoveWindow("Webcam", 1280, 0)
    im = get_image(camera, width, height)
    cv.ShowImage("Webcam", im)
    time.sleep(0.2)

    s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    connected = False
    while not connected:
        try:
            s.connect("\0vidsocket")
            connected=True
        except:
            pass

    try:
        datapath = (s.recv(1024)).decode('latin-1')
        s.send(b'ready')
    except:
        pass

    s.setblocking(0)
    writer = None
    nframes = 0
    time0 = time.time()
    time1 = time0
    while True:
        time1 = time.time()
        # print "Precision:", (time1-time0) - nframes*target_dur, "ms",
        datadec = ""
        try:
            data = s.recv(1024)
            datadec = data.decode('latin-1')
            has_data = True
        except:
            has_data = False

        im = get_image(camera, width, height, writer, s)

        cv.ShowImage("Webcam", im)
        capture_dur = time.time()-time1
        rem = target_dur-capture_dur
        tol = 1.0*1e-3
        if rem > tol:
            nframes += 1
        else:
            # Drop frames until we're in sync again:
            dropped_frames = int((capture_dur+tol)/target_dur)
            # print "Dropped",  dropped_frames, "frames, ",
            nframes += dropped_frames+1
            for nf in range(dropped_frames):
                if writer is not None:
                    write_send(camera, im, writer, s)
            
        next_timestamp = (nframes * target_dur) + time0
        total_wait = next_timestamp - time.time()

        # cv.WaitKey will block with 0:
        while total_wait < tol:
            total_wait += target_dur
            nframes += 1
            if writer is not None:
                write_send(camera, im, writer, s)

        # print "will wait for", total_wait*1e3 , "ms to synchronize",
        key = cv.WaitKey(int(1000.0*total_wait))
        while key != -1:
            next_timestamp = (nframes * target_dur) + time0
            if time.time() < next_timestamp:
                total_wait = next_timestamp - time.time()
            else:
                break
            key = cv.WaitKey(int(1000.0*total_wait))

        if True:
            # No update from blender in a long time, terminate process
            if (has_data and datadec.find('1') == -1 and datadec.find('avi') == -1):# or (not has_data):
                if connected:
                    t_disconnect = time.time()
                connected = False
                if time.time()-t_disconnect > 0.1:
                    sys.stdout.write('WEBCAM: Received termination signal... ')
                    sys.stdout.flush()
                    break
            else:
                connected=True

            # Stop the recording
            if has_data and datadec.find('stop') != -1 and writer is not None:
                print("WEBCAM: Stopping video")
                del(writer)
                writer = None
                # cmd = ['a2mp4.sh', '%s' % fn, '%s.mp4' % fn[:-4]]
                # proc = subprocess.Popen(cmd, bufsize=-1)
            if has_data and datadec.find('avi') != -1 and datadec.find('stop') == -1:
                if writer is None:
                    print("WEBCAM: Starting video")
                    fn = datadec[datadec.find("begin")+5:datadec.find("end")]
                    writer = cv.CreateVideoWriter(fn, cv.CV_FOURCC('X','V','I','D'), fps, (width,height), True)
                connected=True

        # except:
        #     pass

        # print

    
    cv.DestroyWindow("Webcam")    
    s.close()
    sys.stdout.write("done\n")
    sys.exit(0)